### PR TITLE
Add lando warnings support

### PR DESCRIPTION
**Pull request by abpostelnicu**
With https://github.com/mozilla-conduit/lando-api/pull/172 we have support in lando to send warnings.
We are going to send two warnings for each diff for each revision:

- Once the revision has been updated/created an initial message will be send indicating the user that there are linters running:"Review time linters are still running. Are you sure you want to continue?"
- Just as the review bot finishes posting issues in phab we send the 2nd message with the issues or if there aren't any we clear previous send message.

### CODE REVIEW

**Comment by abpostelnicu** [comment]
This has been thoroughly tested on testing and it works like a charm. Unfortunately repo `mozilla-central` is not compatible with the new feature from lando and the warning messages cannot be displayed in the UI bu they appear in the backend.

**Comment by abpostelnicu** [comment]
@zzzeid do you think we can add `mozilla-central` to be compatible on lando testing with the new feature set?

File: bot/code_review_bot/workflow.py [hunk 1]
```diff
@@ -79,6 +80,9 @@ def __init__(
         # Setup Backend API client
         self.backend_api = BackendAPI()
 
+        # Setting up lando
+        self.lando = lando
```
Comment (marco-c) [inline bot/code_review_bot/workflow.py:84]: Nit: `lando_api`

File: bot/code_review_bot/workflow.py [hunk 2]
```diff
@@ -191,6 +197,41 @@ def publish(self, revision, issues, task_failures, links):
         )
         stats.add_metric("analysis.issues.publishable", nb_publishable)
 
+        if self.lando is not None:
+            logger.info(
+                "Publishing warnings to lando for {0} errors and {1} warnings".format(
+                    nb_publishable_errors, nb_publishable - nb_publishable_errors
```
Comment (marco-c) [inline bot/code_review_bot/workflow.py:203]: Nit: replace `nb_publishable - nb_publishable_errors` with `nb_publishable_warnings`

File: bot/code_review_bot/workflow.py [hunk 3]
```diff
@@ -191,6 +197,41 @@ def publish(self, revision, issues, task_failures, links):
...
+                ),
+                revision=revision.id,
+                diff=revision.diff["id"],
+            )
+            self.lando.del_all_warnings(revision.id, revision.diff["id"])
```
Comment (marco-c) [inline bot/code_review_bot/workflow.py:208]: Can you add a comment explaining that we delete all warnings here to remove the warning related to the build being in progress?

File: bot/code_review_bot/workflow.py [hunk 4]
```diff
@@ -191,6 +197,41 @@ def publish(self, revision, issues, task_failures, links):
...
+                    + (
+                        " {} warnings.".format(nb_publishable_warnings)
+                        if nb_publishable_warnings
+                        else ""
+                    )
+                    + " Are you sure you want to continue?"
+                )
+                self.lando.add_warning(
+                    lando_warning_message, revision.id, revision.diff["id"]
+                )
```
Comment (marco-c) [inline bot/code_review_bot/workflow.py:233]: Instead of having this code here, it'd be cleaner to build a new `Reporter` class (like we have the `PhabricatorReporter`), e.g. `LandoReporter`.

Comment (abpostelnicu) [reply bot/code_review_bot/workflow.py:233]: I thought about that, but we would have a special case when no matter if we have or don't have issues pending we will need to do reporter.publish()

File: bot/code_review_bot/workflow.py [hunk 5]
```diff
@@ -191,6 +197,41 @@ def publish(self, revision, issues, task_failures, links):
...
+                    nb_publishable_errors, nb_publishable - nb_publishable_errors
+                ),
+                revision=revision.id,
+                diff=revision.diff["id"],
+            )
+            self.lando.del_all_warnings(revision.id, revision.diff["id"])
+
+            # Determine if we need to post something or not
+            if nb_publishable > 0:
+                lando_warning_message = (
```
Comment (marco-c) [inline bot/code_review_bot/workflow.py:212]: Nit: Let's define the template at the top of the file, so it's more readable (like https://github.com/mozilla/code-review/blob/ef9459f2211cd9eca869e5ea60faecb508288bc1/bot/code_review_bot/report/mail.py#L17).

We should also make it clear as part of the message that errors will likely lead to a back-out.
We could write something like:
`The code review bot found XXX errors which should be fixed to avoid backout and XXX warnings`

File: tools/code_review_tools/lando.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# Heroku Documentation: https://devcenter.heroku.com/
```
Comment (marco-c) [inline tools/code_review_tools/lando.py:6]: Nit: this comment line is unrelated to this file

File: tools/code_review_tools/lando.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
```
Comment (marco-c) [inline tools/code_review_tools/lando.py:1]: Could you move this file to https://github.com/mozilla/libmozdata instead of adding it here?
We might want to use it in other projects too.

File: events/code_review_events/workflow.py [hunk 8]
```diff
@@ -32,21 +32,26 @@
...
+from code_review_tools.lando import LandoWarnings
 
 logger = structlog.get_logger(__name__)
 
 PULSE_TASK_GROUP_RESOLVED = "exchange/taskcluster-queue/v1/task-group-resolved"
 PULSE_TASK_COMPLETED = "exchange/taskcluster-queue/v1/task-completed"
 PULSE_TASK_FAILED = "exchange/taskcluster-queue/v1/task-failed"
 
+LANDO_WARNING_MESSAGE = (
+    "Review time linters are still running. Are you sure you want to continue?"
```
Comment (marco-c) [inline events/code_review_events/workflow.py:44]: `Static analysis and linting is still in progress` (no need for the second part "Are you sure you want to continue?" as this part is already in the Lando UI itself)

File: bot/code_review_bot/workflow.py (hunk 4 above)
Comment (marco-c) [reply bot/code_review_bot/workflow.py:233]: Yes, but that's not a problem, the PhabricatorReporter also does that check

**Comment by marco-c** [comment]
Could you write some tests for this feature in `bot` and `events`?

File: tools/code_review_tools/lando.py (hunk 7 above)
Comment (abpostelnicu) [reply tools/code_review_tools/lando.py:1]: OK, I will move it to libmozdada

File: bot/code_review_bot/report/lando.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            if nb_publishable > 0:
+                self.lando_api.add_warning(
+                    LANDO_MESSAGE.format(
+                        errors=nb_publishable_errors, warnings=nb_publishable_warnings
+                    ),
+                    revision.id,
+                    revision.diff["id"],
+                )
+        except Exception as ex:
+            logger.info(str(ex))
```
Comment (marco-c) [inline bot/code_review_bot/report/lando.py:65]: In this case we should log an error rather than an info message

File: bot/code_review_bot/report/lando.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        logger.info(
+            "Publishing warnings to lando for {0} errors and {1} warnings".format(
+                nb_publishable_errors, nb_publishable - nb_publishable_errors
+            ),
+            revision=revision.id,
+            diff=revision.diff["id"],
+        )
+
+        try:
+            # code-review.events sends an initial warning message to lando,
```
Comment (marco-c) [inline bot/code_review_bot/report/lando.py:52]:
```suggestion
            # code-review.events sends an initial warning message to lando to specify that the analysis is in progress,
```

File: bot/code_review_bot/workflow.py [hunk 11]
```diff
@@ -183,6 +183,7 @@ def publish(self, revision, issues, task_failures, links):
         nb_publishable_errors = sum(
             1 for i in issues if i.is_publishable() and i.level == Level.Error
         )
+
```
Comment (marco-c) [inline bot/code_review_bot/workflow.py:186]: Nit: please remove this extraneous newline

File: events/code_review_events/workflow.py [hunk 12]
```diff
@@ -134,6 +145,20 @@ async def process_build(self, build):
...
+                    "Begin publishing init warning message to lando.",
+                    revision=build.revision["id"],
+                    diff=build.diff_id,
+                )
+                try:
+                    self.lando_warnings.add_warning(
+                        LANDO_WARNING_MESSAGE, build.revision["id"], build.diff_id
+                    )
+                except Exception as ex:
+                    logger.info(str(ex))
```
Comment (marco-c) [inline events/code_review_events/workflow.py:160]: Here too, we should log an error

**Comment by marco-c** [comment]
Looks good overall, but we need to add some tests to both `bot` and `events`

File: bot/code_review_bot/report/lando.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import structlog
+
+from code_review_bot import Level
+from code_review_bot.report.base import Reporter
+
+logger = structlog.get_logger(__name__)
+
+LANDO_MESSAGE = "The code review bot found {errors} errors which should be fixed to avoid backout and {warnings} warnings."
```
Comment (marco-c) [inline bot/code_review_bot/report/lando.py:13]: Nit: we could use the singular form when there is a single error or a single warning

File: bot/tests/test_lando.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
```
Comment (marco-c) [inline bot/tests/test_lando.py:1]: Can you rename this file `test_reporter_lando.py`? This way we match the names of the other tests for the other reporters.

File: events/code_review_events/workflow.py [hunk 15]
```diff
@@ -39,14 +40,30 @@
 PULSE_TASK_COMPLETED = "exchange/taskcluster-queue/v1/task-completed"
 PULSE_TASK_FAILED = "exchange/taskcluster-queue/v1/task-failed"
 
+LANDO_WARNING_MESSAGE = "Static analysis and linting is still in progress."
```
Comment (marco-c) [inline events/code_review_events/workflow.py:43]: Should it be
```suggestion
LANDO_WARNING_MESSAGE = "Static analysis and linting are still in progress."
```
?

I'm not sure which one is grammatically correct

File: events/code_review_events/workflow.py [hunk 16]
```diff
@@ -39,14 +40,30 @@
...
+LANDO_FAILURE_MESSAGE = (
+    "Static analysis and linting did not run due to general failure."
```
Comment (marco-c) [inline events/code_review_events/workflow.py:45]:
```suggestion
    "Static analysis and linting did not run due to a generic failure."
```

File: events/tests/test_workflow.py [hunk 17]
```diff
@@ -48,7 +80,9 @@ async def test_blacklist(PhabricatorMock, mock_taskcluster):
 
 
 @pytest.mark.asyncio
-async def test_publish_results_success_mode(PhabricatorMock, mock_taskcluster):
+async def test_publish_results_success_and_error_mode(
```
Comment (marco-c) [inline events/tests/test_workflow.py:83]: Let's split this into multiple smaller tests instead of modifying the existing test, this way we know right away, in case of failure, what exactly is failing

Comment (abpostelnicu) [reply events/tests/test_workflow.py:83]: Cool I'm going to do 3 tests, lando on success, fail and general_fail.

**Review by marco-c — APPROVED** [review]
